File: 0761-special-binary-string/0761-special-binary-string.py
class Solution:
    def makeLargestSpecial(self, s: str) -> str:
        l = 0
        res = []
        cnt = 0
        for r, c in enumerate(s):
            cnt += 1 if c == "1" else -1
            if cnt == 0:
                inner = self.makeLargestSpecial(s[l+1:r])
                res.append("1" + inner + "0")
                cnt = 0
                l = r + 1

        return "".join(sorted(res, reverse=True))

        

File: 0761-special-binary-string/0761-special-binary-string.java
class Solution {

    public String makeLargestSpecial(String s) {
        /**
        
        1 1 0 0 1 0 1 0 1 1 1 0 0 0
        
         */

        int cnt = 0, l = 0;
        List<String> specialSubs = new ArrayList<>();

        for (int r = 0; r < s.length(); r++) {
            int add = s.charAt(r) == '1' ? 1 : -1;
            cnt += add;
            if (cnt == 0) {
                specialSubs.add("1" + makeLargestSpecial(s.substring(l+1, r)) + "0");
                cnt = 0;
                l = r + 1;
            }
        }

        StringBuilder sb = new StringBuilder();
        specialSubs.sort(Comparator.reverseOrder());
        for (String x: specialSubs) {
            sb.append(x);
        }
        return sb.toString();        
    }
}